macros/PGbasicmacros.pl

```diff
@@ -2128,6 +2141,45 @@ sub EV3P {
     %{$option_ref},
   );
   my $string = join(" ",@_);
+
+  # Preprocessing code to track internal balance (parity) of certain begin/end formatting
+  # variables by putting in calls to the Perl functions which track the status/counts for
+  # the internalBalancing sanity checks in a preprocess phase
+  # Note: This is being done after default_preprocess_code() from lib/WeBWorK/PG/Translator.pm
+  #       replaced "\\" by "\\\\" so we need 4 backslashes before the start/end braces below.
+
+    $string =~ s/\$BBOLD(\W)/\$BBOLD \\\\{ internalBalancingIncrement("openBold"); \\\\}\1/g;
+    $string =~ s/\$EBOLD(\W)/\$EBOLD \\\\{ internalBalancingDecrement("openBold"); \\\\}\1/g;
+
+    $string =~ s/\$BITALIC(\W)/\$BITALIC \\\\{ internalBalancingIncrement("openItalic"); \\\\}\1/g;
+    $string =~ s/\$EITALIC(\W)/\$EITALIC \\\\{ internalBalancingDecrement("openItalic"); \\\\}\1/g;
+
+    $string =~ s/\$BUL(\W)/\$BUL \\\\{ internalBalancingIncrement("openUnderline"); \\\\}\1/g;
+    $string =~ s/\$EUL(\W)/\$EUL \\\\{ internalBalancingDecrement("openUnderline"); \\\\}\1/g;
+
+    $string =~ s/\$BCENTER(\W)/\$BCENTER \\\\{ internalBalancingIncrement("openCenter"); \\\\}\1/g;
+    $string =~ s/\$ECENTER(\W)/\$ECENTER \\\\{ internalBalancingDecrement("openCenter"); \\\\}\1/g;
+
+    $string =~ s/\$BM(\W)/\$BM \\\\{ internalBalancingTurnOn("inInlineMath"); \\\\}\1/g;
+    $string =~ s/\$EM(\W)/\$EM \\\\{ internalBalancingTurnOff("inInlineMath"); \\\\}\1/g;
+
+    $string =~ s/\$BDM(\W)/\$BDM \\\\{ internalBalancingTurnOn("inDisplayMath"); \\\\}\1/g;
+    $string =~ s/\$EDM(\W)/\$EDM \\\\{ internalBalancingTurnOff("inDisplayMath"); \\\\}\1/g;
+
+    $string =~ s/\$BEGIN_ONE_COLUMN(\W)/\$BEGIN_ONE_COLUMN \\\\{ internalBalancingTurnOn("inOneColumnMode"); \\\\}\1/g;
+    $string =~ s/\$END_ONE_COLUMN(\W)/\$END_ONE_COLUMN \\\\{ internalBalancingTurnOff("inOneColumnMode"); \\\\}\1/g;
+
+    $string =~ s/\$BLABEL(\W)/\$BLABEL \\\\{ internalBalancingTurnOn("inInputLabel"); \\\\}\1/g;
+    $string =~ s/\$ELABEL(\W)/\$ELABEL \\\\{ internalBalancingTurnOff("inInputLabel"); \\\\}\1/g;
+
+    # The following two lines are related to new formatting variables added to development
+    # versions of PG by https://github.com/openwebwork/pg/pull/323 which use an HTML span to
+    # allow LTR text inside an RTL context.
+    $string =~ s/\$BLTR(\W)/\$BLTR \\\\{ internalBalancingIncrement("openSpan"); \\\\}\1/g;
+    $string =~ s/\$ELTR(\W)/\$ELTR \\\\{ internalBalancingDecrement("openSpan"); \\\\}\1/g;
+
+  # End preprocessing for internalBalancing sanity checks
+
```

Comment:
All the other function in `EV3P()` are controlled by the `%options` hash above.  I'd recommend addition an option for this as well, so that it can be turned on or off as needed.  You can initialize it from the `warn_on_internalBalance_errors` environment variable if you want.  I would also move these out to a separate function (like `EV3P_parser()` below).

I'm also wondering if you could use a zero-width lookahead pattern `(?=\W)` rather than `(\W)`, as that would not require replacing `\1` in the replacement (apparently parenthesis sub matching is expensive).  This will also allow you to match two variables in a row, e.g., `$BBOLD$BITALIC`, which your current form will not match (because the first match will be for `$BBOLD$`, and after the replacement, the rest of the string is `ITALIC` and so won't match your pattern).  That means your current version might not properly count some situations, and will report errors incorrectly.

Reply:
@dvpc - Thanks. Any additional feedback is welcome.

The new commits address most of this:
  * Added the stack (to track open tags), and using it to detect out-of-order issues and verify valid nesting.
  * Moved the reg-exp string replacement processing from EV3P() into a new subroutine called EV3P_parity_preprocessor().
  * Run the preprocessing only when a conditional passes, now using the value of `run_parity_preprocessor` which has been added to the options hash.
    * An earlier commit from today tested the $warn_on_internalBalance_errors variable directly, but as you said, it may be convenient to control this behavior using the options hash.
  * Changed to the zero-width look-ahead pattern, as suggested. 
    * Thanks for pointing out this technique, I have not seen it in the past.
    * I think the only cases the old code would have missed were when the same tag appeared twice in a row like $BBOLD$BBOLD as each tag is processed in a different regexp replacement line.

Notes:
  * For now I left in the use of status variables/counters:
    * It could be that there is no longer a real need for them.
    * However, it seems a simple method to save and use the current status to detect nesting of things which are not expected to nest (one-column mode, inline math mode, display math mode, being inside an input label).
    * It also detects balance errors without scanning the stack, which becomes necessary if case of improper nesting.
  * It seems that the 4 special functions could be replaced by 2: a "Push" and a "Pop" function which could depend on the type "inFoo" and "openFoo" to adjust some behavior based on whether nested use is expected or not. However, I think that 4 simple functions are probably easier to understand than 2 more complicated functions.
  * Some of the "mess" in the special functions is related to checking that the "type" is supported and tracked, in case someone would try to use these functions directly with a non-tracked type.

Reply:
> I think the only cases the old code would have missed were when the same tag appeared twice in a row like `$BBOLD$BBOLD` as each tag is processed in a different regexp replacement line.

Right.  Sorry, I had been considering asking the the regular expressions be combined, but decided against it.  I had that in mind, I guess.

> it seems a simple method to save and use the current status to detect nesting of things which are not expected to nest (one-column mode, inline math mode, display math mode, being inside an input label).

I'm not sure what you mean for inline and display math modes.  These are processed as a chunk (not as separate begin and end values), and not using variable substitution, so I'm not sure how this would work.

---

I have three additional comments:

1.  Your code should not be run when both `processCommands` and `processVariables` are both true.  If `processCommands` is false, then you parity commands will not be processed, and you will be inserting text that will not be removed (so will be changing the output), and if `processVariables` is false, then there it is incorrect to balance these values.  So the condition should include all three options.

2.  Your replacement string introduced extra spaces into the output text (between the variable and the `\{...\}`).  These may be significant, for example in `$BBOLD something$EBOLD,` would produce "**something** ," (with a space between the `something` and the `,`.  This not only is not what was written, it also introduces unwanted breakpoints, so you could get "**something**" on one line, with the comma at the beginning of the next line.

3.  Finally, your regular expression doesn't take braces into account.  I often use `This is ${BBOLD}bold$EBOLD.` when the variable must be used without a space following it.  For this example, you would miss the `${BBOLD}` and would report a balance error with the `$EBOLD`.  (Usually I will brace both of them, but this example is perfectly legal).  So I think the regular expression needs to be more complicated (sigh).

Reply:
[The following is from an email exchange, posted here to keep the history in place.]

> I think I understand what is needed in terms of items 2 and 3. I think it may be simpler to add an extra regexp with the braces then to create one which can handle the case of with and without braces at once. I'll need to do some review from the Camel books and some Googling+reading. 

One could do separate braced and unbraced versions, but I don't think thats necessary.  You could use

	/\$BBOLD(?=\W)/

to

	/\$(?:BBOLD|\{BBOLD\})(?=\W)/

which seems not too bad to me, and probably better than two separate searches.

> I also noticed that EV3P merges the array of "lines" it is sent into one single long string just before the location where I put in the call to the new EV3P_parity_preprocessor.  It may make sense do all the preprocessing to each line in the array, one at a time, before the lines are merged into one long string. The each line can be tested for matches to a "group" before running the regexp replacements on it (when needed). Ex. testing for say BOLD before running the BOLD regexp replacements on a given line. I'm not sure if this would be more efficient or not, but it seems likely that it should be more efficient, as "many" lines may not need any processing, and can be skipped after a single search rather than several searches by the various regexp operations. Any thought on this?

I have not done any testing, but I think you will find that doing single line expressions will be less efficient.  The regex engine is quite efficient even on large pieces of text, especially for simple patterns that are mostly runs of text, like yours.  As I recall, the regular expression string search complexity is linear in the size of the string used, which means it is roughly the same search time to run the replacement once on the combined strings as it would be to run it separately on the individual strings; but there is a setup cost per substitution (setting up the string to be searched, as I recall) of O(2^m), where m is the size of the search expression, so doing multiple (single-line) searches is more expensive than doing one combined search.

I also don't think that doing a pre-search for BOLD is the way to go.  You are trading off an extra search initially against the two searches for the begin-end pair if it does match.  So when there is NO match, you win, but when there IS a match, you end up with an extra search.  I'm not sure it is worth the extra effort.

You can reduce the number of searches by using an executable replacement via something like

	s/(\$([BE])(BOLD))(?=\W)/EV3P_begin_end($2, $3, $1)/e;

with something like (untested):

	sub EV3P_begin_end {
	  my ($be, $type, $var) = @_;
	  return ($be eq 'B' ? 
	    "$var\\{internalBalancingIncrement('$type');\\}" :
	    "$var\\{internalBalancingDecrement('$type');\\}");
 	};

Of course, you would have to change the defined types to be "BOLD", etc, rather than "openBold" and the others that you are using (or you could have a hash that translated from BOLD to openBold, and so on).  This lets you do both the begin and end checking in one substitution, which makes the checking for BOLD unnecessary (since this substitution already does that).

You could even go so far as to do something like

	s/(\$(?:([BE])(BOLD|ITALIC|UL|CENTER|D?M|LABEL|LTR)|(BEGIN|END)_(ONE_COLUMN)))/EV3P_begin_end($2||$4, $3||$5, $1)/e;

and

	sub EV3P_begin_end {
	  my ($be, $type, $var) = @_;
	  return ($be eq 'B' || $be eq 'BEGIN' ? 
	    "$var\\{internalBalancingIncrement('$type');\\}" :
	    "$var\\{internalBalancingDecrement('$type');\\}");
 	};

This would let you handle all the begin/end elements in one substitution, which would be more efficient (even though the regex is more complicated).  Because all are in one search, the EV3P_begin_end() function will be called for each variable in order, and so you can actually do the balancing right in that routine itself, rather than inserting commands to be executed later.  For example

	sub EV3P_begin_end {
	  my ($be, $type, $var) = @_;
	  if ($be eq 'B' || $be eq 'BEGIN') {
	    internalBalancingIncrement($type);
	  } else {
	    internalBalancingDecrement($type);
	  }
	  return $var;
 	};

This means the string is never actually changed (there replacement string is the same as the original), but the balancing functions are called at the time the substitution is processed.  So more efficient that inserting commands to be processed individually later.  Also, you would no longer need to make the processing dependent on the processCommands option.

You can add support for the braces as well.  Something like:

	$EV3P_pattern = qr/(?:([BE])(BOLD|ITALIC|UL|CENTER|D?M|LABEL|LTR)|(BEGIN|END)_(ONE_COLUMN))/;

	s/(\$(?:\{$EV3P_pattern\}|$EV3P_pattern))(?=\W)/EV3P_begin_end($2||$4||$6||$8, $3||$5||$7||$9, $1)/eo;

with the EV3P_begin_end() directly above.  Of course, this uses a lot of capturing parentheses, but it could be done with only one:

	$EV3P_pattern = qr/(?:[BE](?:BOLD|ITALIC|UL|CENTER|D?M|LABEL|LTR)|(?:BEGIN|END)_ONE_COLUMN)/;

	s/(\$(?:\{$EV3P_pattern\}|$EV3P_pattern))(?=\W)/EV3P_begin_end($1)/eo;

	sub EV3P_begin_end {
	  my $match = shift;
	  my $type = $match;
	  $type =~ s/[${}]//g;                  # remove dollar sign and braces
	  my $be = substr($type,0,1);           # get first character
	  $type =~ s/^(?:BEGIN_|B|END_|E)//;    # remove begin/end character or string
	  if ($be eq 'B') {
	    internalBalancingIncrement($type);
	  } else {
	    internalBalancingDecrement($type);
	  }
	  return $match;
	}

If you use a hash to turn the $type into your original categories, you can simplify this:

	%EV3P_balance_items = (
	  '$BBOLD' => 'openBold',
	  '$EBOLD' => 'openBold',
	  '$BITALICS' => 'openItalics',
	  '$EITALICS' => 'openItalics',
	  ...
	);

	sub EV3P_begin_end {
	  my $match = shift;
	  my $type = $EV3P_balance_items{$match};
	  if ($match =~ m/^\$\{?B/) {
	    internalBalancingIncrement($type);
	  } else {
	    internalBalancingDecrement($type);
	  }
	  return $match;
	}

I think something like this will be more efficient than what you are currently doing or suggesting.


> I'm less sure that I understood item 1, and think my confusion may really just be that a small text correction is needed in your feedback comment. You wrote:
> 
>> Your code **should not be run** when both processCommands and processVariablesare both true. If processCommands is false, then you parity commands will not be processed, and you will be inserting text that will not be removed (so will be changing the output), and if processVariables is false, then there it is incorrect to balance these values. So the condition should include all three options.
> 
> I think your intention was that the new code **should only be run** if both these flags are true (in addition to the flag added just for this code), as either one of the flags you mentioned is false - some part of the intended processing will fail. Is that correct?

Yes, you are correct.  I had edited the line several times, and apparently left it in a bad state.  I've corrected the comment on line.

> About "inline" and "display" math, ... in the WW context, I really mean to address $BM $EM vs $BDM $EDM  which are macros in macros/PGbasicmacros.pl defined by the lines below.

I had forgotten about these (if I even knew about them), and have never seen them used in practice.  I searched the OPL and found only one problem that used them, so I think they are not worth nothing about.  New problems are unlikely to be written using them.

My comment was about the fact that \(...\) and \[...\] are processed as units, so balancing them as you are other variables isn't necessary.  But I forgot about $BM/$EM and $BDM/$EDM.

---

One more thing to consider:  this balancing really only needs to be done during the creation or editing of a problem, not all the time.  There should probably be a development mode that is enabled by the PG editor (and perhaps the library browser), and your processing would only be done when that is in effect.  Once the problem has been developed (and is correct), there is no need to do this balancing every time a student views the problem or submits an answer.  That is a waste of resources for no reason.